### PR TITLE
Attendance FLS check + empty state

### PR DESCRIPTION
**Pull request by bethbrains**
# Critical Changes

# Changes
- Check that running user has all required FLS to conduct Attendance.
- Provide a helpful message in the Attendance component when no Service Participants exist for a given Service Session. 

# Issues Closed

# New Metadata

# Deleted Metadata

# Definition of Done
  Refer to [Asteroids DoD document](https://salesforce.quip.com/iq2mAy4i62oM) to see any additional details for the items below
- [ ] Place holder data is incorporated (Refer to [Case Management Sample Data document](https://quip.com/cbFcAPJF8t0z))
- [x] Base axe-core JEST test included for any new LWC (No critical violations)
- [x] CRUD/FLS is enforced in Apex
- [x] Permission sets & field sets are updated to account for CRUD/FLS/TABS for new object metadata
- [x] All modified classes are unit tested (Apex) at 100% coverage
- [x] UX approval or UX not necessary
- [x] PR contains draft release notes
- [x] Attach work item to the pull request with [Lurch](https://salesforce.quip.com/50ZRA5LEWVzH) `**Lurch:attach W-000000` or `**Lurch:add`
- [ ] All **acceptance criteria** have been met
    - [x] Developer
    - [ ] Code Reviewer
    - [ ] QA
- [ ] QE story level testing completed

### CODE REVIEW

**Review by lbeller — APPROVED** [review]

**Comment by bethbrains** [comment]
[W-8430959](https://gus.lightning.force.com/lightning/r/ADM_Work__c/a07B0000008it5gIAA/view)

File: force-app/main/default/classes/AttendanceService.cls [hunk 1]
```diff
@@ -12,6 +12,8 @@ public with sharing class AttendanceService {
     private AttendanceSelector attendanceSelector = new AttendanceSelector();
     @TestVisible
     private ServiceDeliveryDomain domain = new ServiceDeliveryDomain();
+    @TestVisible
+    private PermissionValidator permValidator = new PermissionValidator();
```
Comment (jjbennett) [inline force-app/main/default/classes/AttendanceService.cls:16]: To be consistent should we set the instance on the perm validator?

File: force-app/main/default/classes/AttendanceService.cls [hunk 2]
```diff
@@ -42,6 +44,75 @@ public with sharing class AttendanceService {
         }
     }
 
+    public Map<String, Boolean> checkFieldPermissions() {
+        Map<String, Boolean> result = new Map<String, Boolean>();
```
Comment (jjbennett) [inline force-app/main/default/classes/AttendanceService.cls:48]: What do you think about adding a method to PermissionValidator that takes a list of SObjectFields, we are converting to string then back to describe. It might be a little easier to read if we just pass a list of fields?

File: force-app/main/default/lwc/attendance/__tests__/attendance.test.js [hunk 3]
```diff
@@ -21,11 +22,13 @@ const mockGenerateRoster = require("./data/generateRoster.json");
 const mockGenerateRosterEmpty = require("../../__tests__/data/emptyList.json");
 const mockGetFieldSet = require("./data/getFieldSet.json");
 const mockWiredSession = require("./data/wiredSession.json");
+const mockWiredPermissions = require("./data/wiredPermissions.json");
```
Comment (jjbennett) [inline force-app/main/default/lwc/attendance/__tests__/attendance.test.js:25]: What about adding a couple new tests with the empty state and when perms return false?

File: force-app/main/default/lwc/emptyState/emptyState.js-meta.xml [hunk 4]
```diff
@@ -0,0 +1,5 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<LightningComponentBundle xmlns="http://soap.sforce.com/2006/04/metadata">
+    <apiVersion>48.0</apiVersion>
```
Comment (jjbennett) [inline force-app/main/default/lwc/emptyState/emptyState.js-meta.xml:3]: Should we up this to the latest?

File: force-app/main/default/lwc/attendance/attendance.js [hunk 5]
```diff
@@ -159,6 +183,25 @@ export default class Attendance extends LightningElement {
...
     }
 
+    get hasPermissions() {
+        return this.isReadOnly
+            ? this.hasReadPermissions
+            : this.hasReadPermissions && this.hasWritePermissions;
+    }
+
+    get showUpdateButton() {
+        return this.isReadOnly && this.hasWritePermissions;
```
Comment (jjbennett) [inline force-app/main/default/lwc/attendance/attendance.js:193]: I wonder if we should change the variable name from isReadOnly to isReadMode or something? Reading this line was a little confusing for me at first.

**Review by jjbennett — APPROVED** [review]
Looking spiffy! I need to take it for a spin yet in the UI but I wanted to get my initial thoughts back to you.

**Review by jjbennett — CHANGES_REQUESTED** [review]
After discussing in slack, we will make the perm validator more robust.

**Review by lbeller — APPROVED** [review]

File: force-app/main/default/classes/AttendanceService_TEST.cls [hunk 6]
```diff
@@ -324,6 +325,32 @@ public with sharing class AttendanceService_TEST {
...
+
+        Test.startTest();
+        Map<String, Boolean> actual = service.checkFieldPermissions();
+        Test.stopTest();
+
+        permValidatorStub.assertCalled(
+            methodName,
+            new List<Type>{ List<Schema.SObjectField>.class, String.class }
+        );
+        System.assertEquals(expected, actual);
```
Comment (jjbennett) [inline force-app/main/default/classes/AttendanceService_TEST.cls:351]: I think the results of the expected vs actual will give us enough information if this where to fail but I know we also talked about adding a message to all asserts. I will leave it up to you.

File: force-app/main/default/classes/PermissionValidator_TEST.cls [hunk 7]
```diff
@@ -102,15 +102,55 @@ private class PermissionValidator_TEST {
...
+    @IsTest
+    private static void testHasFLSAccessInsertNonCreatableFieldByField() {
+        Boolean nonCreatableField = PermissionValidator.getInstance()
+            .hasFLSAccessForFields(Contact.LastModifiedById, 'insert');
+        System.debug(JSON.serialize(Contact.LastModifiedById.getDescribe()));
+        System.assert(!nonCreatableField, 'Can not insert system field, expect false');
+    }
+
+    @IsTest
+    private static void testHasFLSAccessUpsertNonReparentableByField() {
```
Comment (jjbennett) [inline force-app/main/default/classes/PermissionValidator_TEST.cls:125]: Is the name for this correct or should it be SystemField?

**Review by jjbennett — APPROVED** [review]
I really like the new lwc tests, it is almost like specs :) - Fantastic work!

**Review by lbeller — APPROVED** [review]

**Review by jjbennett — APPROVED** [review]

**Review by jjbennett — APPROVED** [review]

**Review by lbeller — APPROVED** [review]
